unittest/server/src/ivi_controller_uinttests.cpp

```diff
@@ -0,0 +1,3913 @@
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ****************************************************************************/
+
+#include <gtest/gtest.h>
+#include <dlfcn.h>
+#include <string>
+#include "server_api_fake.h"
+#include "ivi_layout_interface_fake.h"
+#include "ivi_layout_structure.hpp"
+
+enum {
+    WESTON_BUFFER_SHM,
+    WESTON_BUFFER_DMABUF,
+    WESTON_BUFFER_RENDERER_OPAQUE,
+    WESTON_BUFFER_SOLID,
+} type;
+
+extern "C"
+{
+#include "ivi-controller.c"
+}
+
+const struct wl_interface wl_buffer_interface, wl_output_interface;
+static constexpr uint8_t OBJECT_NUMBER = 2;
+static uint32_t g_SurfaceCreatedCount = 0;
+
+struct wl_resource * custom_wl_resource_from_link(struct wl_list *link)
+{
+	struct wl_resource *resource;
+
+	return wl_container_of(link, resource, link);
+}
+
+struct wl_list * custom_wl_resource_get_link(struct wl_resource *resource)
+{
+	return &resource->link;
+}
+
+static uint32_t custom_get_id_of_surface(struct ivi_layout_surface *ivisurf)
+{
+    return ivisurf->id_surface;
+}
+
+static uint32_t custom_get_id_of_layer(struct ivi_layout_layer *ivilayer)
+{
+    return ivilayer->id_layer;
+}
+
+static int32_t custom_get_layers_on_screen(struct weston_output *output, int32_t *pLength,
+					                       struct ivi_layout_layer ***ppArray)
+{
+    struct ivi_layout_layer ivilayer;
+    ivilayer.id_layer = 1;
+
+    *pLength = 1;
+    *ppArray = (ivi_layout_layer**)malloc(sizeof(struct ivi_layout_layer *));
+    (*ppArray)[0] = &ivilayer;
+
+    return 0;
+}
+
+static int32_t custom_get_surfaces_on_layer(struct ivi_layout_layer *ivilayer, int32_t *pLength,
+				                            struct ivi_layout_surface ***ppArray)
+{
+    struct ivi_layout_surface ivisurface;
+    ivisurface.id_surface = 1;
+
+    *pLength = 1;
+    *ppArray = (ivi_layout_surface**)malloc(sizeof(struct ivi_layout_surface *));
+    (*ppArray)[0] = &ivisurface;
+
+    return 0;
+}
+
+static int32_t custom_surface_get_size(struct ivi_layout_surface *ivisurf, int32_t *width, int32_t *height,
+			                           int32_t *stride)
+{
+    *width = 1;
+    *height = 1;
+    *stride = 1;
+
+    return 0;
+}
+
+static int custom_read_pixels(struct weston_output *output,
+			                  pixman_format_code_t format, void *pixels,
+			                  uint32_t x, uint32_t y,
+			                  uint32_t width, uint32_t height)
+{
+    return x == 0 || y == 0;
+}
+
+static void surface_create_callback(struct wl_listener *listener, void *data)
+{
+    g_SurfaceCreatedCount++;
+}
+
+class ControllerTests : public ::testing::Test
```

Comment:
I think this function should only aim for SetUp() and TearDown(). The controller content should be a separate class.

Reply:
we will consider to do this thing in refactor action in the future